Add support for CreateDomain action

jakubcolony opened the pull request:
## Description

This PR adds support for the CreateDomain action.

<img width="995" alt="image" src="https://user-images.githubusercontent.com/112586815/220772541-2fe9eb52-c0e7-4cb5-9abd-0877dbff98e2.png">

## Testing
1. Create a colony using the wizard.
2. Click on the Test Create Domain action.
3. You'll be redirected to a transaction not found page, wait a few seconds and refresh.
4. You should see the action details (including domain name, color, description). Please verify the domain is added to the domain list dropdown as well.

Resolves #279

jakubcolony commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -134,22 +134,22 @@ enum Network {
 
 enum DomainColor {
-  LIGHTPINK
+  LIGHT_PINK
```
Changes those to have properly generated PascalCase members in CDapp

rdig replied:
You mean the underscores ?

jakubcolony replied:
Yeah, if you don't include those, you get `Lightpink` instead of `LightPink`

jakubcolony commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -297,14 +297,21 @@ type User @model {
-  description: String
-  color: DomainColor
-  parent: Domain @hasOne
+  isRoot: Boolean!
```
I thought this is an easier solution since we only support one level nesting (both in Dapp and on the contract level). Another thing that pushed me towards that change was that I couldn't get the parentId from the `DomainAdded` event or by calling `getDomain` on colony client - something to look at when we want to start supporting more complex nesting

rdig replied:
Sure, I'm good with that.

Just keep in mind that at some point we'll need to deal with multiple nesting levels

jakubcolony commented on amplify/backend/api/colonycdapp/schema.graphql:
```diff
@@ -297,14 +297,21 @@ type User @model {
+  metadata: DomainMetadata @hasOne(fields: ["id"])
+}
+
+type DomainMetadata @model {
+  id: ID! # This field is referenced by Domain id, so has to be in the same format: colonyAddress_nativeId
+  name: String!
+  description: String!
+  color: DomainColor!
```
Separated this since it's a piece of information that's only available in the CDapp (and doesn't get emitted to the network, therefore inaccessible in the block-ingestor). This way, the saga is only responsible for creating the metadata, and the actual domain entity is created by action handler in block-ingestor

The advantage this gives is that block-ingestor will still pick up the events that originated from a contract call as opposed to a saga inside CDapp. Those will not have any metadata attached, but will still show correctly in the actions list.

We can follow the same pattern in the colony creation saga

jakubcolony commented on package.json:
```diff
@@ -33,7 +33,7 @@
   "homepage": "https://colony.io",
   "devDependencies": {
-    "@aws-amplify/cli": "^10.6.1",
+    "@aws-amplify/cli": "10.4.1",
```
https://github.com/aws-amplify/amplify-category-api/issues/1014 😠

jakubcolony commented on src/components/common/ColonyActions/ColonyActions.tsx:
```diff
@@ -206,6 +207,22 @@ const ColonyActions = (/* { ethDomainId }: Props */) => {
         )}
         text="Test Create Payment"
       />
+      <ActionButton
+        submit={ActionTypes.ACTION_DOMAIN_CREATE}
+        error={ActionTypes.ACTION_DOMAIN_CREATE_ERROR}
+        success={ActionTypes.ACTION_DOMAIN_CREATE_SUCCESS}
+        transform={pipe(
+          mergePayload({
+            colonyAddress: colony.colonyAddress,
+            colonyName: colony.name,
+            domainName: 'Test Domain',
+            domainColor: DomainColor.BlueGrey,
+            domainPurpose: 'Test Domain Description',
+          }),
+          withMeta({ navigate }),
+        )}
+        text="Test Create Domain"
+      />
```
Heads up: this is for testing only

jakubcolony commented on src/types/domains.ts:
```diff
@@ -1,57 +1,20 @@
 import { DomainColor } from '~gql';
 
-export enum Color {
```
I think we don't need two separate enums (and a mapping between them) and we can simply use the GraphQL enum directly

rdig replied:
Works for me.

rdig commented on src/components/shared/DomainDropdown/DomainDropdown.tsx:
```diff
@@ -90,7 +89,16 @@ const DomainDropdown = ({
-          domain={ALLDOMAINS_DOMAIN_SELECTION}
+          domain={{
+            id: '',
+            isRoot: true,
```
This is a bit misleading. 

"All teams" include all domains, including root. See it more of a grouping of domains, rather than a nested tree/list.

So maybe leave a comment explaining this so future devs won't be confused

jakubcolony replied:
This is actually a dummy domain data that we use to display the "All Teams" option in the dropdown. `isRoot` has to be set to true so that it's not shown as a subdomain item like so:
<img width="342" alt="image" src="https://user-images.githubusercontent.com/112586815/221062102-17d48183-e63d-4aab-8c31-143378e451e5.png">

But I agree it's confusing without further explanation so I added a relevant comment

rdig reviewed (approved):
Looks good. Left some comments, but otherwise stellar work

jakubcolony commented:
@rdig A couple of extra changes since your approval:
- Modified temp-create-data and colonyCreate saga to call the domain metadata mutation and create the root domain in the database manually (since it doesn't trigger an event)
- Removed createUniqueDomain lambda function

Feel free to have another look when you have time

Julianahlk reviewed (approved):
Working well for me and looks good on mobile 👍

willm30 commented on src/components/shared/DomainDropdown/DomainDropdown.tsx:
```diff
@@ -118,7 +130,8 @@ const DomainDropdown = ({
         .filter(notNull)
```
The `.slice()` and the comment re subgraph can presumably be removed now

willm30 commented on src/components/shared/DomainDropdown/DomainDropdown.tsx:
```diff
@@ -118,7 +130,8 @@ const DomainDropdown = ({
         .sort(sortByDomainId)
         .map((domain) => {
-          const { nativeId, name: domainName } = domain || {};
+          const { nativeId, metadata } = domain || {};
```
We expect `domain` to be defined here since we filter `notNull` above

willm30 reviewed (approved):
Nice one 👍
